### PR TITLE
Make replication informer async

### PR DESCRIPTION
**Pull request by danielmarbach**
Part of #11543

### CODE REVIEW

File: Raven.Client.Lightweight/Connection/ReplicationInformer.cs [hunk 1]
```diff
@@ -44,22 +44,29 @@ public ReplicationInformer(DocumentConvention conventions, HttpJsonRequestFactor
...
                 if (lastReplicationUpdate.Add(Conventions.TimeToWaitBetweenReplicationTopologyUpdates) > SystemTime.UtcNow)
-                    return new CompletedTask();
+                    return;
             }
-            lock (replicationLock)
+
+            try
             {
+                // Do a quick synchronous check before we resort to async/await with the state-machine overhead.
+                if (!replicationLock.Wait(0))
```
Comment (danielmarbach) [inline Raven.Client.Lightweight/Connection/ReplicationInformer.cs:68]: Previously this acquired a replication lock object while `RefreshReplicationInformationInternal` acquired `lock(this)` so I assumed it is OK to assume only one operation is ever allowed into this critical section. Please advise

File: Raven.Client.Lightweight/Connection/ReplicationInformer.cs [hunk 2]
```diff
@@ -72,20 +79,22 @@ private Task UpdateReplicationInformationIfNeededInternalAsync(string url, Func<
...
 
-                var taskCopy = refreshReplicationInformationTask;
-                if (taskCopy != null)
-                    return taskCopy;
+                if (refreshReplicationInformationTask != null)
+                    await refreshReplicationInformationTask.ConfigureAwait(false);
 
-                return refreshReplicationInformationTask = Task.Factory.StartNew(() => 
-                    RefreshReplicationInformationInternal(url, getReplicationDestinations)).
+                refreshReplicationInformationTask = RefreshReplicationInformationInternalAsync(url, getReplicationDestinations).
```
Comment (danielmarbach) [inline Raven.Client.Lightweight/Connection/ReplicationInformer.cs:102]: If I understood the code correctly the rationale is to do the heavy lifting in the background but keep track of it so I kept that consistent

File: Raven.Client.Lightweight/Connection/ReplicationInformer.cs [hunk 3]
```diff
@@ -72,20 +79,22 @@ private Task UpdateReplicationInformationIfNeededInternalAsync(string url, Func<
...
 
                 if (lastReplicationUpdate.Add(Conventions.TimeToWaitBetweenReplicationTopologyUpdates) > SystemTime.UtcNow && force == false)
-                    return new CompletedTask();
+                    return;
 
-                var taskCopy = refreshReplicationInformationTask;
-                if (taskCopy != null)
-                    return taskCopy;
+                if (refreshReplicationInformationTask != null)
+                    await refreshReplicationInformationTask.ConfigureAwait(false);
```
Comment (danielmarbach) [inline Raven.Client.Lightweight/Connection/ReplicationInformer.cs]: If I understood correctly next refresh should await the previous refresh task if available and then return to not trigger subsequent refreshes

File: Raven.Client.Lightweight/Connection/ReplicationInformer.cs [hunk 4]
```diff
@@ -140,26 +155,34 @@ protected override string GetServerCheckUrl(string baseUrl)
...
-            RefreshReplicationInformationInternal(serverClient.Url, () => AsyncHelpers.RunSync(() => serverClient.DirectGetReplicationDestinationsAsync(new OperationMetadata(serverClient.Url, serverClient.PrimaryCredentials, null), null)));
+            AsyncHelpers.RunSync(() =>
+                RefreshReplicationInformationInternalAsync(serverClient.Url, () => serverClient.DirectGetReplicationDestinationsAsync(new OperationMetadata(serverClient.Url, serverClient.PrimaryCredentials, null), null)));
         }
 
         public override void RefreshReplicationInformation(ServerClient serverClient)
         {
-            RefreshReplicationInformationInternal(serverClient.Url, () => serverClient.DirectGetReplicationDestinations(new OperationMetadata(serverClient.Url, serverClient.PrimaryCredentials, null), null));
+            AsyncHelpers.RunSync(() =>
+                RefreshReplicationInformationInternalAsync(serverClient.Url, () => Task.FromResult(serverClient.DirectGetReplicationDestinations(new OperationMetadata(serverClient.Url, serverClient.PrimaryCredentials, null), null))));
```
Comment (danielmarbach) [inline Raven.Client.Lightweight/Connection/ReplicationInformer.cs:180]: Wasn't entirely sure here whether you would prefer to offload `serverClient.DirectGetReplicationDestinations` or not. Given the sync nature of this code I went straight to not offload. Please advise

File: Raven.Client.Lightweight/Connection/ReplicationInformer.cs (hunk 1 above)
Comment (ayende) [reply Raven.Client.Lightweight/Connection/ReplicationInformer.cs:68]: Yes

File: Raven.Client.Lightweight/Connection/ReplicationInformer.cs (hunk 2 above)
Comment (ayende) [reply Raven.Client.Lightweight/Connection/ReplicationInformer.cs:102]: Yes, looks okay

**Review by ayende — APPROVED** [review]

**Comment by danielmarbach** [comment]
By the way, this has slightly more allocation costs due to more async machinery involved. Yet I think this is negligible in comparison to the previously sync over async problem this code had. In addition to that, I removed the LINQ overhead in this call to compensate allocs. That being said I haven't measured the before and after nor would I know how to properly isolate this to measure other than doing long-running profiling sessions which would be a significant effort on my side.

**Review by grisha-kotler — APPROVED** [review]

**Comment by danielmarbach** [comment]
Rebased

@grisha-kotler @ayende @ppekrol anything that hinders this from being merged?

**Comment by ppekrol** [comment]
Hi @danielmarbach 

We have scheduled 5 CI builds and all of them got stuck. We are looking into that at the moment.

**Comment by ayende** [comment]
Looks like we are hanging here: 
image

**Comment by ayende** [comment]
This is pretty consistently getting stuck there. I took a brief look and it looks like the `refreshReplicationInformationTask` is never completed. It is possible that this is related to `SemaphoreSlim` usage, but I'm not going to dive too deeply into this.

This command reproduce this quite easily:  

` C:\Work\ravendb-3.5\Raven.Tests> ..\SharedLibs\xunit\xunit.console.clr4.exe .\bin\Debug\Raven.Tests.dll`

I don't see much benefit in this change at this time. This is happening on an infrequent basis and shouldn't have much of an impact.

**Comment by ayende** [comment]
You also need to handle disposed cancellation token, but after fixing that, we have hangs here. 
I don't think that this is a change that we want to push forward, given the critical nature of the location and the impact.


```
 	Raven.Abstractions.dll!Raven.Abstractions.Util.AsyncHelpers.ExclusiveSynchronizationContext.BeginMessageLoop() Line 153	C#
 	Raven.Abstractions.dll!Raven.Abstractions.Util.AsyncHelpers.RunSync(System.Func<System.Threading.Tasks.Task> task) Line 45	C#
 	Raven.Client.Lightweight.dll!Raven.Client.Connection.AdminServerClient.EnsureDatabaseExists(string name, bool ignoreFailures) Line 38	C#
 	Raven.Client.Lightweight.dll!Raven.Client.Document.DocumentStore.Initialize(bool ensureDatabaseExists) Line 472	C#
 	Raven.Tests.Helpers.dll!Raven.Tests.Helpers.RavenTestBase.NewRemoteDocumentStore(bool fiddler, Raven.Server.RavenDbServer ravenDbServer, string databaseName, bool runInMemory, string dataDirectory, string requestedStorage, bool enableAuthentication, bool ensureDatabaseExists, System.Action<Raven.Client.Document.DocumentStore> configureStore, string activeBundles, System.Collections.Generic.IEnumerable<Raven.Client.Indexes.AbstractIndexCreationTask> indexes, System.Collections.Generic.IEnumerable<Raven.Client.Indexes.AbstractTransformerCreationTask> transformers, System.Collections.Generic.IEnumerable<System.Collections.IEnumerable> seedData, bool noStaleQueries) Line 358	C#
>	Raven.Tests.dll!Raven.Tests.Faceted.LazyFacets.Default_operator_not_honoured_remote_store_ToFacetsLazy() Line 22	C#

```

**Comment by danielmarbach** [comment]
Fair enough

**Comment by danielmarbach** [comment]
I do think though some of those hangs could have been mitigated opting out from context capturing with

```
            using (NoSynchronizationContext.Scope())
            {
```

Pretty sure the reason of these hangs is XUnit custom context.

**Pull request closed without merging**